Add design of watcher for IoT use cases

manoranjith opened the pull request:
Relates to perun-proposal #003.

Signed-off-by: Manoranjith <ponraj.manoranjitha@in.bosch.com>

matthiasgeihs commented:
The proposal is a bit hard to parse.

I think it would be a good start with a rough overview of the suggested watcher functionality:

- The watcher is responsible for tracking the on-chain state in acting on behalf of the user if the user is offline.
- The user always pushes the latest channel state to the watcher. requires `Watcher.PushState`
- The watcher watches the chain for state registrations. requires `OnChain.EventSubscription`
- If an older state is registered on-chain, the watcher automatically disputes the state with the newer state. requires `OnChain.Dispute`

Next, it would be helpful to have the proposed types and interfaces written down more explicitly and compactly, like

```go
type Watcher struct {...}

func NewWatcher(rs RegisterSubscriber) *Watcher {...}

func (w *Watcher) Watch(ChannelID, StateSub, StatePub) {...}
func (w *Watcher) StopWatch(ChannelID) {...}
func (w *Watcher) Shutdown(ChannelID) {...}

type StateSub interface {
	Next() Event
}

type StatePub inteface {
	Push(ChannelState)
}

type RegistererSubscriber interface {
	Registerer
	Subscriber
}

type Registerer interface {
	Register(ChannelState)
}

type Subscribe interface {
	Subscribe(ChannelID)
}
```

matthiasgeihs commented:
Another thing to consider:

How does the watcher handle sub-channels and virtual channels? Currently we watch sub-channels individually. However, we will soon change our logic so that only full channel trees can be registered. This also means that sub-channels cannot be registered without their parent channels. Hence, it should be sufficient to watch the root ledger channel. It also means that the watcher needs the subchannel states in order to register a parent channel.

manoranjith commented:
@matthiasgeihs I have revised the draft proposal as per your suggestion (starting the description with an overview of functionalities, followed by a concise description of interfaces).

I have also added a section that explains how different types of channels (ledger channel, sub-channel and virtual channel) should be handled by the watcher. I have written this based on how sub-channels and virtual channels are implemented in the latest go-perun release (v0.7.0).

Can you review the proposal again ?

danielksan81 commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+}
+
+// For receiving registered and progressed events for a given channel ID.
+type Subscribe interface {
```
Is there a typo? Should this be Subscribe**r** instead of Subscribe?

manoranjith replied:
Yes, it should be `Subscriber`. Have fixed it in efb39d7.

choeppler commented:
NB: I renamed the base branch from master to main to align with Hyperledger policy. Cf. https://lists.hyperledger.org/g/labs/message/145. The first attempt didn't work, that's why there is a note on me having changed the base branch from main to master above...

matthiasgeihs commented:
@manoranjith Did you think of how a hub watches its virtual channels? See [LINK](https://github.com/hyperledger-labs/go-perun/blob/749b197450cf4e82e302e5ff2552a3ee744e265e/client/virtual_channel.go#L101). The hub does only look for on-chain registration and always stores the newest registered state. It does not become active in a refutation process as it is not really interested in which state is concluded. It is only important to the hub that the concluded state is the same for both participants, which is always true if the participants are on the same network and use the same smart contract. (It is not necessarily true if the participants are on different networks.)

manoranjith commented:
> @manoranjith Did you think of how a hub watches its virtual channels? See [LINK](https://github.com/hyperledger-labs/go-perun/blob/749b197450cf4e82e302e5ff2552a3ee744e265e/client/virtual_channel.go#L101). The hub does only look for on-chain registration and always stores the newest registered state. It does not become active in a refutation process as it is not really interested in which state is concluded. It is only important to the hub that the concluded state is the same for both participants, which is always true if the participants are on the same network and use the same smart contract. (It is not necessarily true if the participants are on different networks.)

Hi @matthias,

I tried to go over this scenario and interestingly the refutation for virtual channel will work when participants are on same network and use same smart contracts. Here's what will happen:

### Setup of Ledger Channels  (`LC1` and `LC2`)
1. Alice and Ingrid would have established `LC1` and funded it on-chain. They start watching individually for events on `LC1`.
2. Same is the case for ledger channel `LC2` between Bob and Ingrid.

### Setup of Virtual Channel (`VC`)
1. Alice and Bob agree on initial state of virtual channel.
2. Alice sends to Ingrid, a funding update on `LC1` with initial state of `VC`
3. Ingrid waits for an update from Bob as well.
4. Bob sends to Ingrid, a funding update on `LC2` with initial state of `VC`
5. Ingrid checks if both the initial states are matching.
6. If yes, it accepts both states and virtual channel is funded.
7. Now Alice, Ingrid, Bob have the initial state of `VC` and  can individually start watching for events on `VC`.

### On-Chain Dispute Registered on `VC`
1. Alice and Bob make many off-chain transactions. Say, latest version of `VC` is 10.
2. Alice turns out to be malicious, so registers LC1 with an older state of `VC` (say version 9) on the blockchain.

### Scenario 1: Bob never responds.
1. Let's assume Bob's watcher is offline for some reason and does not respond to this registered event on `VC`.
2. Now, since Ingrid is not aware of any state on `VC` with version greater than 9, Ingrid also does nothing and the channel `LC1` is settled.
3. After `LC1` is settled, Bob comes online and wants to settle `VC` with version 10.
4. Since the version 10 state of VC is not finalized, it cannot settle `LC2` with Ingrid via off-chain update.
5. Even if the version 10 state of `VC` was finalized and Bob sends a withdrawal update to Ingrid, Ingrid can identify the state of `VC` in `LC2` is different from that in `LC1` (which was setteld) and hence reject the update.
6. Now, if Bob registers `LC2` along with `VC` (with version 10) on blockchain, this call will fail. Because
    1.  `VC` was already registered with a different version and the refutation timeout for it has passed.
7. So, the only option for Bob is to settle with `VC` version 9, with through off-chain update on `LC2` or by registering on-chain dispute for `LC2`.

The above scenario shows that, when a dispute is registered for a virtual channel by Alice (using ledger channel `LC1`), Bob **must** refute with the latest state of the virtual channel before the challenge duration expires.

### Scenario 2
1. On receiving the event that `VC` was registered with version 9, Bob will refute with the latest state (version 10).
2. Since latest state for VC has been registered, the challenge duration will expire and the `VC` can be concluded.

Thus, we see that, hub (Ingrid) need not be aware of the updates on the virtual channel. Even if the virtual channel is registered for on-chain dispute, the two participants (Alice and Bob) will ensure that latest state of the virtual channel is registered on the blockhain.

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+
+// For receiving on-chain events from the watcher.
+type OnChainEventsSub interface {
+	Next() OnChainEvent
```
Do you define type `OnChainEvent`?

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+
+// For receiving off-chain states from the main component.
+type OffChainStatesSub interface {
+	Next() OffChainState
```
Do you define type `OffChainState`?

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
````diff
@@ -0,0 +1,248 @@
+
+
+```go
+type Watcher struct{}
````
```suggestion
type Watcher struct{ ... }
```

manoranjith replied:
Updated.

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+}
+
+// For sending off-chain states to the watcher.
+type OffChainStatesPub inteface {
```
```suggestion
type OffChainStatesPub interface {
```

matthiasgeihs replied:
Where and how is this used?

manoranjith replied:
Updated.

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+// Interfaces for the main component to interact with the watcher.
+
+// For receiving on-chain events from the watcher.
+type OnChainEventsSub interface {
```
Where and how is this used?

manoranjith replied:
Have updated comments for this and all other interfaces to describe their usage.

manoranjith replied:
I have re-ordered the sections by moving interface definition after the interaction diagram. Because, the interface definition is derived from the interaction diagram.

Do you think it is easy to relate the two now ?

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+Interaction between the watcher and main component based using the above
+interfaces is shown in the diagram below.
+
+![Interaction between the main component and watcher](004/watcher.svg)
```
Can you insert the function names defined above? Otherwise, it may not be clear what you refer to.

The first step should be to create the watcher?

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+<!-- New Feature (Introduces a functionality) -->
+Architecture (Requires a modification of the architecture)
+
+## Implementation
```
I am missing details on what needs to be changed in go-perun to integrate the watcher concept.

manoranjith replied:
I have briefly described the changes required in go-perun in points 1 and 2 of implementation. These are the only two changes, I foresee now. After these changes, I think the watcher could be implemented. 

Do you think, we should add a more detailed description of changes required ?

After the watcher is implemented, it would be upto the user of go-perun framework to initialize the watcher and register channels to it. On the other hand, we could also provide an option to make go-perun internally start and manage the watcher. What do you think ?

matthiasgeihs reviewed (commented):
First review, focusing on high level concepts.

I am missing a connection between the diagram and the interface definition. How are the interfaces used?

manoranjith commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+// For registering off-chain state for a given ledger channel,
+// along with the off-chain states of its sub-channels and virtual channels.
+type Registerer interface {
+    Register(State, SubStates)
```
I have added the return values wherever it is necessary.

manoranjith commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+
+// For receiving registered and progressed events for a given channel ID.
+type Subscriber interface {
+    Subscribe(ChannelID)
```
Yes, we need it here. I have added the return values wherever it is necessary.

manoranjith commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+type Watcher struct{}
+
+// For initalizing the watcher.
+func NewWatcher(rs RegisterSubscriber) *Watcher {...}
```
I have re-organized the interface definitions into 3 code blocks.

manoranjith commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+// For initalizing the watcher.
+func NewWatcher(rs RegisterSubscriber) *Watcher {...}
+
+// For interacting with the blockchain.
```
Updated.

manoranjith commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,248 @@
+}
+
+// For registering/de-registering a channel ID with/from the watcher.
+func (w *Watcher) Register(ChannelID, OffChainStatesSub, OnChainEventsPub) {...}
```
Yes, I have added error returns wherever necessary.

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -0,0 +1,363 @@
+
+// AdjudicatorEventsPub is used by the watcher to send notifications of
+// adjudicator events to the main component.
+type AdjudicatorEventsPub inteface {
```
```suggestion
type AdjudicatorEventsPub interface {
```

matthiasgeihs reviewed (commented):
after chatting, i thought you asked me to do another review. now i see that you did not indicate that here, so i am not sure. anyways, since i spent some time looking at your comment replies, here are my thoughts:

Some of my comments are not addressed yet, among them, maybe the [most important one](https://github.com/hyperledger-labs/perun-proposals/pull/11#discussion_r690355806):

I have trouble seeing how the interfaces would be used by the client. I think it would help a lot if you create a short code sketch that shows how the interfaces would be used by a client. For example, like the following, where I already hint some open issues in the comments:

```
w, err := NewWatcher(rs) \\ client creates watcher. why does the client need to provide RegisterSubscriber? Couldn't that be an implementation detail of the watcher?

ch := ... // client opens channel
err := w.StartWatching(c.ID, stateSub, eventSub) // client instructs watcher to watch channel. Why does the client need to provide stateSub and eventSub? Could the function create those for the client and return them?

... // client updates channel
err = stateSub.Publish(c.state) // client informs watcher about new channel state.

.. // client closes channel
err = w.StopWatching(c.ID) // client informs watcher that the channel no longer needs to be watcher.
err = w.Shutdown() // client shuts down watcher at end of program. does the client really need to shut down the watcher?
```

manoranjith commented:
> after chatting, i thought you asked me to do another review. now i see that you did not indicate that here, so i am not sure. anyways, since i spent some time looking at your comment replies, here are my thoughts:
> 
> Some of my comments are not addressed yet, among them, maybe the [most important one](https://github.com/hyperledger-labs/perun-proposals/pull/11#discussion_r690355806):
> 
> I have trouble seeing how the interfaces would be used by the client. I think it would help a lot if you create a short code sketch that shows how the interfaces would be used by a client. For example, like the following, where I already hint some open issues in the comments:
> 
> ```
> w, err := NewWatcher(rs) \\ client creates watcher. why does the client need to provide RegisterSubscriber? Couldn't that be an implementation detail of the watcher?
> 
> ch := ... // client opens channel
> err := w.StartWatching(c.ID, stateSub, eventSub) // client instructs watcher to watch channel. Why does the client need to provide stateSub and eventSub? Could the function create those for the client and return them?
> 
> ... // client updates channel
> err = stateSub.Publish(c.state) // client informs watcher about new channel state.
> 
> .. // client closes channel
> err = w.StopWatching(c.ID) // client informs watcher that the channel no longer needs to be watcher.
> err = w.Shutdown() // client shuts down watcher at end of program. does the client really need to shut down the watcher?
> ```

@matthiasgeihs  Yes, I did ask you to do another review. On

#### 1. Including exact names in the sequence diagram:
I first designed the interaction and based on that, designed the interfaces. So, I thought of not including, as mentioned it in [this comment](https://github.com/hyperledger-labs/perun-proposals/pull/11#discussion_r691215656). But, I had misplaced this response on a different thread, which caused the confusion.

#### 2. Why the client should provide `RegistererSubscriber` interface to the `Watcher`. 

I think, I mixed up two distinct steps:
    1. Initializing the watcher component.
    2. Connecting to the watcher component from client.
 
This could be the cause of confusion. Will update the proposal to clarify this.

### 3. Why the pub-sub interfaces are passed by the main component to the watcher instead of other way round:

Because, this way the main component is in control of the pub-sub interface. So, if at a later point of time;
1. If the main component wants to close this watcher and use a different one or,
2. If the main component wants to use an additional watcher,

It will be easier to do, if the main component is in control of the pub-sub instance. I have already included this explanation in the proposal [(here)](https://github.com/boschresearch/perun-proposals/blob/004-iot-adoption-watcher/design/004-IoT-Adoption-Watcher.md#note-on-pub-sub-interfaces-between-the-main-component-and-the-watcher).

#### 4. Including a snippet indicate usage

1. This would definitely help. Will add it to the proposal.

manoranjith commented:
@matthiasgeihs I have addressed all the review comments:
1. Used exact names in the interaction diagram (25c1080).
2. Add a section describing illustrative usage of API (603556b).
3. Revised the interface for initializing watcher and include watcher initialization in sequence diagram (eecba5f).
4. Move the reason for initializing the pub-sub instances from description to rationale section (0b15c23).

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
````diff
@@ -308,20 +319,90 @@ discard the older ones.
+#### Using the watcher from the main component
+
+```go
+// Either initialize a local watcher
+w, err := NewWatcher(rs RegistererSubscriber)
+
+// Or connect to the remote watcher
+w, err := ConnectToWatcherGrpc(cfg Config)
````
`local` and `remote` can be two different packages. We can load the RegistererSubscriber in the watcher implementation.
```suggestion
// Either initialize a local watcher
w, err := local.NewWatcher(cfg local.Config)

// Or connect to the remote watcher
w, err := remote.NewWatcher(cfg remote.Config)
```

manoranjith replied:
Updated.

matthiasgeihs commented on design/004-IoT-Adoption-Watcher.md:
```diff
@@ -308,20 +319,90 @@ discard the older ones.
+// 1. Setup the subscription,
+// 2. Instruct the watcher to start watching for events on that channel and,
+// 3. Publish the initial state.
+statesPub, statesSub := newStatesPubSub()
+adjudicatorEventsPub, adjudicatorEventsSub := newAdjudicatorEventsPubSub()
+
+err := w.StartWatching(ch1.ID(), statesSub, adjudicatorEventsPub)
+
+statesPub.Publish(ch.StateWithSignatures())
```
You write that the reason for initializing the pub subs external of the watcher is that you want to be able to pass them to other watchers later.

Another way of supporting concurrent watchers, would be to implement a concurrent watcher type, that internally manages a set of watchers `[w1, w2, ...]`. That is, when you call `w.StartWatching`, it calls `w1.StartWatching`, `w2.StartWatching`, ... . I think it might be more intuitive to use. You could even keep all functionality within the watcher component.

```suggestion
err := w.StartWatching(ch1.ID())
w.PublishState(ch1.ID(), ch1.StateWithSignatures())
e, err := w.NextEvent(ch1.ID())
```

manoranjith replied:
1. I like your suggestion on supporting concurrent watching by implementing a concurrent watcher.
So, I have moved the responsibility of setting up pub-sub interfaces to the watcher. The pub-sub interfaces are now returned by the `StartWatching` API. This is updated in b6a2ddd

2. However, I think we should use `statesPub.Publish(ch.StateWithSignatures())` instead of `w.PublishState(ch1.ID(), ch1.StateWithSignatures())`Because publishing a state is done in the context of the channel and for a given channel, the API should allow publishing states only for that channel.

You could see the last section of updated proposal to see how the API works now.

matthiasgeihs reviewed (commented):
Some more thoughts on the abstract concept. We should discuss in our next call.

manoranjith commented:
@matthiasgeihs As we figured out during our call the other day,

Previously, the focus was on
  1. Designing the internal working of watcher and
  2. Designing the interface between watcher and main component.

But, this was making the scope of the proposal very broad and also causing confusion.
As discussed, I have updated the proposal to focus only on
  
1. Designing the interface between watcher and main component.

Following changes were made to achieve this

  - Remove descriptions of
    - how the watcher consumes the off-chain states published by the main component.
    - how the watcher publishes the adjudicator events to the main component.

  - Move sections on illustrative usage to implementation hints.
  - Add a section on how to integrate the watcher API into go-perun client.
 
See b6a2ddd
